[LT] Add a new backend interface

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76521
* __->__ #76517

Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

Commit:
[LT] Add a new backend interface

Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993480).

alanwaketan commented:
CC @wonjoolee95

Commit:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993480).

wconstab commented on test/cpp/lazy/test_lazy_ops.cpp:
```diff
@@ -21,7 +21,9 @@ namespace lazy {
 namespace {
-  // This registers the torchscript backend, without which lazy device won't work
+// This registers the torchscript backend, without which lazy device won't work.
+// FIXME: This registers the backend for the whole test binary. We should probably
```
We don't currently have a 'de-init' function, do we want to add one?

Does calling init multiple times break anything (should we make that safe, or make it err?)

alanwaketan replied:
Calling init multiple times break things. PyTorch will complain that the lazy backend get re-registered. However, I don't know if we can add a de-init thing. I try to look for APIs that can un-register the lazy backend but failed...

wconstab replied:
well i guess it would be easier to make our init function stateful, so it only really inits once?

alanwaketan replied:
I love the idea. A `static bool` should do.

wconstab reviewed (commented):
This looks fine to me, (nothing looks 'wrong'),

But I'm not too sure if I understand the goal of the change and how it fits into the overall design for 'device'.

Can you write up a little more about the goal in the PR desc?

And can you write up the 'design' for device either in the code somewhere or in the gdoc?
 - what internal/core systems depend on default device (cache, etc)
 - should there be 'default' values in python binding functions, or should user be required to pass in device from python, or should we take device arg out of python bindings (mark step, wait_device_ops, etc)
 - why does SyncLiveTensorsGraph(&backend_device, devices, wait) take both 'backend_device' and 'devices' parameter? is this something we should clean up?

JackCaoG reviewed (approved):
Thanks!

JackCaoG commented:
@wconstab Can we try to merge this pr? I want to start my backend interface experiment from this change.

wconstab reviewed (approved):
oops lost track of this. LGTM. @alanwaketan can you merge?

alanwaketan commented:
> This looks fine to me, (nothing looks 'wrong'),
> 
> But I'm not too sure if I understand the goal of the change and how it fits into the overall design for 'device'.
> 
> Can you write up a little more about the goal in the PR desc?
I hope Jack has explained his needs well.

> 
> And can you write up the 'design' for device either in the code somewhere or in the gdoc?
Sure.

> 
> * what internal/core systems depend on default device (cache, etc)
> * should there be 'default' values in python binding functions, or should user be required to pass in device from python, or should we take device arg out of python bindings (mark step, wait_device_ops, etc)
Fixed.

> * why does SyncLiveTensorsGraph(&backend_device, devices, wait) take both 'backend_device' and 'devices' parameter? is this something we should clean up?
That I don't know. Will need to follow up.

Commit:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

Commit:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

alanwaketan commented:
@JackCaoG Does the XLA failure seem related?

JackCaoG commented:
It seems to be something introcued by this pr
```
2022-05-12T18:20:11.6094789Z ======================================================================
2022-05-12T18:20:11.6095148Z ERROR [0.088s]: test_T_view_xla (__main__.TestViewOpsXLA)
2022-05-12T18:20:11.6095640Z ----------------------------------------------------------------------
2022-05-12T18:20:11.6096029Z Traceback (most recent call last):
2022-05-12T18:20:11.6096732Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 390, in instantiated_test
2022-05-12T18:20:11.6097157Z     raise rte
2022-05-12T18:20:11.6097714Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
2022-05-12T18:20:11.6098126Z     result = test(self, **param_kwargs)
2022-05-12T18:20:11.6098408Z   File "/var/lib/jenkins/workspace/xla/test/../../test/test_view_ops.py", line 607, in test_T_view
2022-05-12T18:20:11.6098660Z     self.assertEqual(t[1, 0], v[0, 1])
2022-05-12T18:20:11.6099085Z   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 627, in assertEqual
2022-05-12T18:20:11.6099356Z     x, y = self.prepare_for_compare(x, y)
2022-05-12T18:20:11.6099631Z   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 570, in prepare_for_compare
2022-05-12T18:20:11.6099933Z     x = tx.to(device='cpu')
2022-05-12T18:20:11.6100156Z RuntimeError: Lazy tensor backend not registered.
```

XLA backend shouldn't reuqire a Lazy tensor backend registration right now..

alanwaketan commented:
> It seems to be something introcued by this pr
> 
> ```
> 2022-05-12T18:20:11.6094789Z ======================================================================
> 2022-05-12T18:20:11.6095148Z ERROR [0.088s]: test_T_view_xla (__main__.TestViewOpsXLA)
> 2022-05-12T18:20:11.6095640Z ----------------------------------------------------------------------
> 2022-05-12T18:20:11.6096029Z Traceback (most recent call last):
> 2022-05-12T18:20:11.6096732Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 390, in instantiated_test
> 2022-05-12T18:20:11.6097157Z     raise rte
> 2022-05-12T18:20:11.6097714Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
> 2022-05-12T18:20:11.6098126Z     result = test(self, **param_kwargs)
> 2022-05-12T18:20:11.6098408Z   File "/var/lib/jenkins/workspace/xla/test/../../test/test_view_ops.py", line 607, in test_T_view
> 2022-05-12T18:20:11.6098660Z     self.assertEqual(t[1, 0], v[0, 1])
> 2022-05-12T18:20:11.6099085Z   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 627, in assertEqual
> 2022-05-12T18:20:11.6099356Z     x, y = self.prepare_for_compare(x, y)
> 2022-05-12T18:20:11.6099631Z   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 570, in prepare_for_compare
> 2022-05-12T18:20:11.6099933Z     x = tx.to(device='cpu')
> 2022-05-12T18:20:11.6100156Z RuntimeError: Lazy tensor backend not registered.
> ```
> 
> XLA backend shouldn't reuqire a Lazy tensor backend registration right now..

So XLA has adopted the BackendDevice interface and possibly some of the BackendInterface?

JackCaoG commented:
so far only `BackendDevice`,  I am working on `BackendInterface`

alanwaketan commented:
> so far only `BackendDevice`, I am working on `BackendInterface`

Yea, that explains. I guess we need to land the backend interface adoption together. Otherwise, it will break XLA.

Commit:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

JackCaoG commented:
Thanks, I will work on a follow up pr on xla side.

Commit:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

Commit:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

Krovatkin committed:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]

Krovatkin committed:
Update on "[LT] Add a new backend interface"


Summary:
This commit introduces a new set of BackendImplInterface: GetDefaultDeviceOrdinal
and SetDefaultDeviceOrdinal. It allows backend to specify their own default
device, e.g, 1 for XLA and 0 for CUDA/CPU.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D35993480](https://our.internmc.facebook.com/intern/diff/D35993480)

[ghstack-poisoned]
